### PR TITLE
Add BF16 Constant Initializer and support for other initializer 

### PR DESCRIPTION
**Pull request by wozna**
### PR types

Others 
### PR changes

Others 
### Describe

This PR: 
- adds BF16 Constant Initializer
- adds support when initializer input is BF16 or `dtype` is set to BF16. Then all initializers except the `Constant Initializer` will initialize the data in FP32 and then ` cast` operator FP32 -> BF16 will be added, so the output from this operator will be BF16.

### CODE REVIEW

**Comment by wozna** [comment]
@arlesniak  @arogowie-intel  Could you please start your review?

**Review by jczaja — DISMISSED** [review]
LGTM

File: paddle/fluid/operators/math/math_function.cc [hunk 1]
```diff
@@ -49,6 +49,7 @@ template struct SetConstant<platform::CPUDeviceContext, platform::complex128>;
 
 #ifdef PADDLE_WITH_XPU
 template struct SetConstant<platform::XPUDeviceContext, platform::float16>;
+template struct SetConstant<platform::XPUDeviceContext, platform::bfloat16>;
```
Comment (arogowie-intel) [inline paddle/fluid/operators/math/math_function.cc:53]: You will need to include header for bfloat16:
```
#include "paddle/fluid/platform/bfloat16.h"
```

File: python/paddle/fluid/tests/unittests/test_fill_constant_op.py [hunk 2]
```diff
@@ -431,5 +431,36 @@ def test_shape_tensor_list_dtype():
...
+        self.dtype = np.uint16
+        self.mkldnn_data_type = "bfloat16"
+
+    def test_check_output(self):
+        self.check_output_with_place(core.CPUPlace())
+
+    def test_error(self):
+        if core.is_compiled_with_cuda():
+            with self.assertRaises(Exception):
+                self.check_output_with_place(core.CUDAPlace(0))
```
Comment (arogowie-intel) [inline python/paddle/fluid/tests/unittests/test_fill_constant_op.py]: Why are you testing the throwing exception and not using 'unittest.skipif' as in our other bf16 tests?

Comment (wozna) [reply python/paddle/fluid/tests/unittests/test_fill_constant_op.py]: I added it to check PADDLE_THROW in `fill_constant_op.h` file, but I removed it in commit https://github.com/PaddlePaddle/Paddle/pull/31935/commits/5cf7140fcd299967e06079440a382579cf5d4e5f. 
I will correct it. .

**Comment by wozna** [comment]
@jczaja @arogowie-intel  Could you please repeat your review?

File: paddle/fluid/operators/math/math_function.cu [hunk 3]
```diff
@@ -33,6 +34,7 @@ using complex64 = paddle::platform::complex64;
 using complex128 = paddle::platform::complex128;
 
 template struct SetConstant<platform::CUDADeviceContext, platform::float16>;
+template struct SetConstant<platform::CUDADeviceContext, platform::bfloat16>;
```
Comment (arlesniak) [inline paddle/fluid/operators/math/math_function.cu:37]: Are that changes needed here?  BF16 is enabled only for CPU.

Comment (wozna) [reply paddle/fluid/operators/math/math_function.cu:37]: Unfortunately, because of fill_constant_op.h implementation it is needed to add this change.
But thanks to https://github.com/PaddlePaddle/Paddle/blob/e2d4ef9d4d8cfb278da3e3860fd503ed639a540c/paddle/fluid/operators/fill_constant_op.h#L109  only the CPU version will be enabled.

File: paddle/fluid/operators/math/math_function.cc (hunk 1 above)
Comment (arogowie-intel) [reply paddle/fluid/operators/math/math_function.cc:53]: That's strange but I do not see added header? Did you add it?

File: python/paddle/fluid/tests/unittests/test_fill_constant_op.py (hunk 2 above)
Comment (arogowie-intel) [reply python/paddle/fluid/tests/unittests/test_fill_constant_op.py]: Unfortunately I do not see changes

**Review by arogowie-intel — APPROVED** [review]

File: paddle/fluid/operators/math/math_function.cu (hunk 3 above)
Comment (wozna) [reply paddle/fluid/operators/math/math_function.cu:37]: To clarify, the point is that in `fill constant op` we can choose the place of calculations (DeviceContext) also using the "place_type" attribute, then in `fill_constant_op.h` we call `math::SetConstant` function where we set DeviceContext and date type. When I registered bfloat16 kernel, during compilation it is checked if math::SetConstant call is supported for all data types.  And I didn't find a better solution not to have to refactor the whole thing.

Comment (arlesniak) [reply paddle/fluid/operators/math/math_function.cu:37]: LGTM

**Review by arlesniak — APPROVED** [review]
LGTM

**Comment by wozna** [comment]
@luotao1 Could you please start your review? 
In PR-CI-CPU-Py2 approval is needed due to changes in `core.ops` where BF16 support has been added.

**Review by luotao1 — APPROVED** [review]

**Review by zhiqiu — APPROVED** [review]
LGTM for paddle.fluid.layers.embedding
